src/api.rs

```diff
@@ -214,41 +225,53 @@ pub fn create_build(
     config: Data<Config>,
     req: HttpRequest,
 ) -> impl Future<Item = HttpResponse, Error = ApiError> {
+    Box::pin(create_build_async(args, db, config, req)).compat()
+}
+
+async fn create_build_async(
+    args: Json<CreateBuildArgs>,
+    db: Data<Db>,
+    config: Data<Config>,
+    req: HttpRequest,
+) -> Result<HttpResponse, ApiError> {
     let repo1 = args.repo.clone();
     let repo2 = args.repo.clone();
-    futures::done(req.has_token_claims("build", "build")).and_then(move |_| {
-        futures::done(req.has_token_repo(&repo1)).and_then(move |_| {
-            futures::done(config.get_repoconfig(&repo2).map(|rc| rc.clone())) // Ensure the repo exists
-                .and_then(move |repoconfig| {
-                    db.new_build(NewBuild {
-                        repo: args.repo.clone(),
-                    })
-                    .and_then(move |build| {
-                        let build_repo_path = config.build_repo_base.join(build.id.to_string());
-                        let upload_path = build_repo_path.join("upload");
-
-                        init_ostree_repo(
-                            &build_repo_path,
-                            &repoconfig.path,
-                            build.id,
-                            &repoconfig.collection_id,
-                        )?;
-                        init_ostree_repo(&upload_path, &repoconfig.path, build.id, &None)?;
-
-                        respond_with_url(&build, &req, "show_build", &[build.id.to_string()])
-                    })
-                })
+
+    req.has_token_claims("build", "build")?;
+    req.has_token_repo(&repo1)?;
+
+    let repoconfig = config.get_repoconfig(&repo2).map(|rc| rc.clone())?; // Ensure the repo exists
+
+    let build = db
+        .new_build(NewBuild {
+            repo: args.repo.clone(),
         })
-    })
+        .await?;
+    let build_repo_path = config.build_repo_base.join(build.id.to_string());
+    let upload_path = build_repo_path.join("upload");
+
+    init_ostree_repo(
+        &build_repo_path,
+        &repoconfig.path,
+        build.id,
+        &repoconfig.collection_id,
+    )?;
+    init_ostree_repo(&upload_path, &repoconfig.path, build.id, &None)?;
+
+    respond_with_url(&build, &req, "show_build", &[build.id.to_string()])
 }
 
 pub fn builds(
     db: Data<Db>,
     req: HttpRequest,
 ) -> impl Future<Item = HttpResponse, Error = ApiError> {
-    futures::done(req.has_token_claims("build", "build"))
-        .and_then(move |_| db.list_builds())
-        .and_then(move |builds| Ok(HttpResponse::Ok().json(builds)))
+    Box::pin(builds_async(db, req)).compat()
```

Comment:
This should probably be marked as async so you no longer need to pin it.

Reply:
I also wonder what is the difference between build & builds_async here?

Reply:
As I understand it, actix-web 1.0 uses futures 0.1, which doesn't support async/await. I haven't figured out how to upgrade everything to actix-web 4.0 yet, so for now this converts the async/await futures to 0.1 futures for actix-web.

I'm not a rust expert, so I'm sure there's a better way to do this somehow.

Reply:
Ah yes, my bad. I missed the compat part somehow.